### PR TITLE
#9 - crud for meetup

### PR DESCRIPTION
**Pull request by MichalMyskow**
I have created basic CRUD system for Meetups:

- models and migration;
- controller and policy for user authentication;
- basic routing for CRUD operations;
- basic views;

### CODE REVIEW

File: src/Http/Controllers/MeetupController.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    public function store(Request $request)
+    {
+        $this->validate($request, [
+            "title" => "required",
+            "date" => "required",
+            "place" => "required",
+            "language" => "required",
+        ]);
```
Comment (kamilpiech97) [inline src/Http/Controllers/MeetupController.php:38]: Move that into form request validation - https://laravel.com/docs/9.x/validation#form-request-validation.

File: src/Http/Controllers/MeetupController.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+        $request->user()->meetups()->create($request->all());
```
Comment (kamilpiech97) [inline src/Http/Controllers/MeetupController.php:40]:
```suggestion
        $request->user()->meetups()->create($request->validated());
```

File: src/Http/Controllers/MeetupController.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        return redirect()->route("meetups");
+    }
+
+    public function edit(Meetup $meetup)
+    {
+        $this->authorize("change", $meetup);
+
+        return view("meetups.edit", [
+            "meetup" => $meetup,
+        ]);
```
Comment (kamilpiech97) [inline src/Http/Controllers/MeetupController.php:51]:
```suggestion
return view("admin.command.admin.index")
           ->with("meetup", $meetup);
```

File: src/Http/Controllers/MeetupController.php [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        $this->authorize("change", $meetup);
+
+        $this->validate($request, [
+            "title" => "required",
+            "date" => "required",
+            "place" => "required",
+            "language" => "required",
+        ]);
+
+        $meetup->update($request->all());
```
Comment (kamilpiech97) [inline src/Http/Controllers/MeetupController.php:65]: The same as above.

File: src/Models/Meetup.php [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        "date",
+        "place",
+        "language",
+    ];
+
+    protected $casts = [
+        "date" => "datetime",
+    ];
+
+    public function user()
```
Comment (kamilpiech97) [inline src/Models/Meetup.php:26]:
```suggestion
    public function user(): BelongsTo
```

Comment (kamilpiech97) [reply src/Models/Meetup.php:26]: remember about types

File: src/Models/User.php [hunk 6]
```diff
@@ -29,4 +29,9 @@ class User extends Authenticatable
     protected $casts = [
         "email_verified_at" => "datetime",
     ];
+
+    public function meetups()
+    {
+        return $this->hasMany(Meetup::class);
+    }
```
Comment (kamilpiech97) [inline src/Models/User.php:36]: The same as above.

File: src/Policies/MeetupPolicy.php [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+declare(strict_types=1);
+
+namespace Blumilk\Meetup\Core\Policies;
+
+use Blumilk\Meetup\Core\Models\Meetup;
+use Blumilk\Meetup\Core\Models\User;
+use Illuminate\Auth\Access\HandlesAuthorization;
+
+class MeetupPolicy
```
Comment (kamilpiech97) [inline src/Policies/MeetupPolicy.php:11]: Do we need policies at the beginning of the project?

File: resources/views/meetups/create.blade.php [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+                <form action="{{ route('meetups') }}" method="post">
+                    @csrf
+                    <div>
+                        <label for="title">Title:</label>
+                        <input type="text" id="title" name="title" value="{{ old('title') }}">
+                        @error('title')
+                        <div>
+                            {{ $message }}
+                        </div>
+                        @enderror
```
Comment (kamilpiech97) [inline resources/views/meetups/create.blade.php:17]: We will think about components on the views.

File: resources/views/meetups/edit.blade.php [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    <div>
+        <div>
+            <h1>Edit Meetup</h1>
+            @auth
+                <form method="post" action="{{ route('meetups.update', $meetup) }}">
+                    @method('PUT')
+                    @csrf
+                    <div>
+                        <label for="title">Title:</label>
+                        <input type="text" id="title" name="title" value="{{ $meetup->title }}">
```
Comment (kamilpiech97) [inline resources/views/meetups/edit.blade.php:13]:
```suggestion
                        <input type="text" id="title" name="title" value="{{ old('title', $meetup->title) }}">
```

Comment (kamilpiech97) [reply resources/views/meetups/edit.blade.php:13]: Please, do like that in other places.

File: resources/views/meetups/edit.blade.php [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+                        <textarea name="description" id="description" cols="30" rows="4">{{ $meetup->description }}</textarea>
+                        @error('description')
+                        <div>
+                            {{ $message }}
+                        </div>
+                        @enderror
+                    </div>
+                    <div>
+                        <label for="date">Date:</label>
+                        <input type="dateTime-local" id="date" name="date" value="{{ $meetup->date->format('Y-m-d\TH:i') }}">
```
Comment (kamilpiech97) [inline resources/views/meetups/edit.blade.php:31]: Move that into model cast.
Like that:
```
protected $casts = [
        "date" => "datetime:Y-m-d h:i:s",
    ];
```

**Review by kamilpiech97 — CHANGES_REQUESTED** [review]

File: src/Http/Controllers/MeetupController.php [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        $this->authorize("change", $meetup);
+
+        return view("meetups.edit", [
+            "meetup" => $meetup,
+        ]);
+    }
+
+    public function update(Request $request, Meetup $meetup)
+    {
+        $this->authorize("change", $meetup);
```
Comment (kamilpiech97) [inline src/Http/Controllers/MeetupController.php:56]: I will move that into routing.

File: src/Http/Controllers/MeetupController.php [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+
+        $this->validate($request, [
+            "title" => "required",
+            "date" => "required",
+            "place" => "required",
+            "language" => "required",
+        ]);
```
Comment (kamilpiech97) [inline src/Http/Controllers/MeetupController.php:63]: The same as above about validation.

File: src/Http/Controllers/MeetupController.php [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+namespace Blumilk\Meetup\Core\Http\Controllers;
+
+use Blumilk\Meetup\Core\Models\Meetup;
+use Illuminate\Http\Request;
+
+class MeetupController extends Controller
+{
+    public function __construct()
+    {
+        $this->middleware(["auth"])->except("index");
```
Comment (kamilpiech97) [inline src/Http/Controllers/MeetupController.php:14]: Move that into routing.
Why except index?

File: src/Http/Controllers/MeetupController.php (hunk 11 above)
Comment (krzysztofrewak) [reply src/Http/Controllers/MeetupController.php:56]: Do you mean "I would move that into routing"?

Comment (kamilpiech97) [reply src/Http/Controllers/MeetupController.php:56]: Yes, I would move that into routing*

File: src/Http/Controllers/MeetupController.php [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    public function index()
+    {
+        $meetups = Meetup::latest()->with(["user"])->paginate(20);
```
Comment (MrHDOLEK) [inline src/Http/Controllers/MeetupController.php:19]: It is not better to transfer such things to the site that the controller is small ?

**Comment by MrHDOLEK** [comment]
@krzysztofrewak Do we write controllers according to the laravel idea or one controller one action?

File: src/Http/Controllers/MeetupController.php (hunk 14 above)
Comment (kamilpiech97) [reply src/Http/Controllers/MeetupController.php:15]: Its small query. That has not a big logic or smth like that. I would stay with that solution.

Comment (MrHDOLEK) [reply src/Http/Controllers/MeetupController.php:15]: But what when there will be a big query for more business logic? It's worth doing the division at the beginning to have order later

Comment (kamilpiech97) [reply src/Http/Controllers/MeetupController.php:15]: For me, that complicates the project at first, but **ok** as you like.

File: src/Http/Requests/StoreMeetupRequest.php [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+
+use Illuminate\Foundation\Http\FormRequest;
+
+class StoreMeetupRequest extends FormRequest
+{
+    public function authorize()
+    {
+        return true;
+    }
+
```
Comment (krzysztofrewak) [inline src/Http/Requests/StoreMeetupRequest.php:15]:
```suggestion
    public function authorize()
    {
        return true;
    }

```

Laravel returns `true` by default from version 8.

File: src/Http/Controllers/MeetupController.php [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+{
+    public function index()
+    {
+        $meetups = Meetup::latest()->with(["user"])->paginate(20);
+
+        return view("meetups.index")
+            ->with("meetups", $meetups);
+    }
+
+    public function create()
```
Comment (krzysztofrewak) [inline src/Http/Controllers/MeetupController.php:21]: Please add return types in all methods. Here it'll be probably `Illuminate\Contracts\View\View`.

File: src/Http/Requests/StoreMeetupRequest.php [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+    public function rules()
```
Comment (krzysztofrewak) [inline src/Http/Requests/StoreMeetupRequest.php:16]:
```suggestion
    public function rules(): array
```

Please change it in other form requests in the same way.

**Review by krzysztofrewak — CHANGES_REQUESTED** [review]

**Review by kamilpiech97 — APPROVED** [review]

**Review by krzysztofrewak — APPROVED** [review]

**Review by MrHDOLEK — APPROVED** [review]

**Comment by krzysztofrewak** [comment]
@MichalMyskow, please merge your first PR. :)

**Pull request merged**